File: MyProject/Content/MyAsset/Lua/MyLua/Libs/EventHandle/EventDispatchGroup.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");
MLoader("MyLua.Libs.DataStruct.MDictionary");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "EventDispatchGroup";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mGroupID2DispatchDic = GlobalNS.new(GlobalNS.MDictionary);
    self.mIsInLoop = false;
end

-- 添加分发器
function M:addEventDispatch(groupID, disp)
    if (not self.mGroupID2DispatchDic:ContainsKey(groupID)) then
        self.mGroupID2DispatchDic:Add(groupID, disp);
    end
end

function M:addEventHandle(groupID, pThis, handle)
    if(not self.mGroupID2DispatchDic:ContainsKey(groupID)) then
        local disp = GlobalNS.new(GlobalNS.EventDispatch);
        self:addEventDispatch(groupID, disp);
    end
    self.mGroupID2DispatchDic:value(groupID):addEventHandle(pThis, handle);
end

function M:removeEventHandle(groupID, pThis, handle)
    if (self.mGroupID2DispatchDic:ContainsKey(groupID)) then
        self.mGroupID2DispatchDic:value(groupID):removeEventHandle(pThis, handle);
    else
        -- 日志
    end
end

function M:dispatchEvent(groupID, dispatchObject)
    self.mIsInLoop = true;
    if (self.mGroupID2DispatchDic:ContainsKey(groupID)) then
        self.mGroupID2DispatchDic:value(groupID):dispatchEvent(dispatchObject);
    else
        -- 日志
    end
    self.mIsInLoop = false;
end

function M:clearAllEventHandle()
    if (not self.mIsInLoop) then
        for _, dispatch in pairs(self.mGroupID2DispatchDic:getData()) do
            dispatch:clearEventHandle();
        end

        self.mGroupID2DispatchDic:Clear();
    else
        -- 日志
    end
end

function M:clearGroupEventHandle(groupID)
    if (not self.mIsInLoop) then
        if self.mGroupID2DispatchDic:ContainsKey(groupID) then
            self.mGroupID2DispatchDic:value(groupID):clearEventHandle();
            self.mGroupID2DispatchDic:Remove(groupID);
        else
            -- 日志
        end
    else
        -- 日志
    end
end

return M;

File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/EventHandle/EventDispatchGroup.cpp
#include "MyProject.h"
#include "EventDispatchGroup.h"
#include "Prequisites.h"
#include "EventDispatch.h"
#include "MClassFactory.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(EventDispatchGroup, GObject)

EventDispatchGroup::EventDispatchGroup()
{
	this->mIsInLoop = false;
}

// 添加分发器
void EventDispatchGroup::addEventDispatch(int groupId, EventDispatch* disp)
{
	if (!this->mGroupId2DispatchDic.containsKey(groupId))
	{
		this->mGroupId2DispatchDic[groupId] = disp;
	}
}

void EventDispatchGroup::addEventHandle(int groupId, EventDispatchDelegate handle)
{
	this->mGroupId2DispatchDic[groupId]->addEventHandle(handle);
}

void EventDispatchGroup::removeEventHandle(int groupId, EventDispatchDelegate handle)
{
	if (this->mGroupId2DispatchDic.containsKey(groupId))
	{
		this->mGroupId2DispatchDic[groupId]->removeEventHandle(handle);
	}
	else
	{
		GLogSys->log("Event Dispatch Group not exist");
	}
}

void EventDispatchGroup::dispatchEvent(int groupId, IDispatchObject* dispatchObject)
{
	this->mIsInLoop = true;

	if (this->mGroupId2DispatchDic.containsKey(groupId))
	{
		this->mGroupId2DispatchDic[groupId]->dispatchEvent(dispatchObject);
	}
	else
	{
		GLogSys->log("Event Dispatch Group not exist");
	}

	this->mIsInLoop = false;
}

void EventDispatchGroup::clearAllEventHandle()
{
	if (!this->mIsInLoop)
	{
		// map for 语句
		for(auto dispatch : this->mGroupId2DispatchDic.getData())
		{
			dispatch.second->clearEventHandle();
		}

		this->mGroupId2DispatchDic.clear();
	}
	else
	{
		GLogSys->log("looping cannot delete element");
	}
}

void EventDispatchGroup::clearGroupEventHandle(int groupId)
{
	if (!this->mIsInLoop)
	{
		if (this->mGroupId2DispatchDic.containsKey(groupId))
		{
			this->mGroupId2DispatchDic[groupId]->clearEventHandle();
			mGroupId2DispatchDic.remove(groupId);
		}
		else
		{
			GLogSys->log("Event Dispatch Group not exist");
		}
	}
	else
	{
		GLogSys->log("looping cannot delete element");
	}
}

bool EventDispatchGroup::hasEventHandle(int groupId)
{
	bool ret = false;

	if (this->mGroupId2DispatchDic.containsKey(groupId))
	{
		ret = this->mGroupId2DispatchDic[groupId]->hasEventHandle();
	}

	return ret;
}

MY_END_NAMESPACE